[appearance: base] Add default `base` properties for checkbox and radio button in UA style sheet

shallawa opened the pull request:
#### e71b8ba98c83ea0d254d3fbd2f8f1a5f8c8476f5
<pre>
[appearance: base] Add default `base` properties for checkbox and radio button in UA style sheet
<a href="https://bugs.webkit.org/show_bug.cgi?id=304092">https://bugs.webkit.org/show_bug.cgi?id=304092</a>
<a href="https://rdar.apple.com/166415728">rdar://166415728</a>

Reviewed by Anne van Kesteren.

-internal-auto-base() can only be parsed in the UA style sheet. Adding conditional
properties to html.css will allow setting some of the CSS properties for checkbox
and radio button when they are styled with &quot;appearance: base&quot;.

Remove all `-internal-auto-base()` properties from the existing layout tests. No
need anymore to enable `CSSInternalAutoBaseParsingEnabled` in these layout tests.

Specs: <a href="https://www.w3.org/TR/css-forms-1/#stylesheet-checkbox-radio">https://www.w3.org/TR/css-forms-1/#stylesheet-checkbox-radio</a>

* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics.html:
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic-expected.html:
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic.html:
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-initial.html:
* Source/WebCore/css/html.css:
(input:is([type=&quot;radio&quot;], [type=&quot;checkbox&quot;])):
(input[type=&quot;radio&quot;]):
(input[type=checkbox]:not([switch]):checked::checkmark):
(input[type=radio]:checked::checkmark):
(input:is([type=checkbox]:not([switch]), [type=radio]):not(:checked)::checkmark):

Canonical link: <a href="https://commits.webkit.org/306540@main">https://commits.webkit.org/306540@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3269fe51a8e06fbff36753ab9edb7c63bf736d8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/141650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/14035 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/3508 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/150222 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/94762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/329b2f50-e2a5-4c6e-84ad-fa8337e2d568) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/143517 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/14746 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/14194 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/108845 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/94762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/33748bb2-e011-4a1c-b299-1b44fba2403d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/144599 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/11387 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/126783 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/89745 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/10941 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/8579 "Passed tests") | [✅ 🛠 wpe-libwebrtc](https://ews-build.webkit.org/#/builders/172/builds/295 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/120226 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/2800 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/152615 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/13725 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/3263 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/116945 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/13740 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/11974 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/117270 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29874 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/13301 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/123487 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/69333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/13763 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/2769 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/13502 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/77488 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/13705 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/13549 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
Can we split this into two commits? One that de-duplicates the existing rules, and one that adds the base properties

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -790,28 +790,34 @@ input:-webkit-autofill, input:-webkit-autofill-strong-password, input:-webkit-au
     margin: 3px 2px;
     box-sizing: border-box;
-    border: initial;
+    border: -internal-auto-base(initial, 0.3em solid currentColor);
```
Doesn't match [the spec](https://drafts.csswg.org/css-forms/#basic-appearance-stylesheet). Does the PR need updating or the spec need updating or is there no decision here yet (in which case, do we have an issue filed)?

shallawa replied:
This was removed.

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -790,28 +790,34 @@ input:-webkit-autofill, input:-webkit-autofill-strong-password, input:-webkit-au
-    border-radius: 5px;
+    border-radius: -internal-auto-base(5px, 0.5em);
+#else
+    border-radius: -internal-auto-base(unset, 0.5em);
```
Doesn't match [the spec](https://drafts.csswg.org/css-forms/#basic-appearance-stylesheet). Does the PR need updating or the spec need updating or is there no decision here yet (in which case, do we have an issue filed)?

shallawa replied:
This was removed.

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -790,28 +790,34 @@ input:-webkit-autofill, input:-webkit-autofill-strong-password, input:-webkit-au
 #endif
 }
 
 input[type="radio"] {
 #if defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY
     border-radius: 50%;
+#else
+    border-radius: -internal-auto-base(unset, 50%);
```
Doesn't match [the spec](https://drafts.csswg.org/css-forms/#basic-appearance-stylesheet). I think it's pretty clear that in this case the spec should be fixed...

fantasai replied:
Filed https://github.com/w3c/csswg-drafts/pull/13227

lukewarlow replied:
This if fixed in the spec now.

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -847,6 +856,31 @@ input:is([type="checkbox"], [type="radio"]):checked:disabled {
 #endif
 }
 
+input[type=checkbox]:not([switch]):checked::checkmark {
+    content: '\2713' / '';
```
It might make more sense to keep this as the content of the checkmark always, and just show/hide it with `visibility`? That would be more performant, I think. But we'd need to file an issue with the spec to match.

fantasai replied:
Filed https://github.com/w3c/csswg-drafts/issues/13226

shallawa replied:
I will address the `visibility` proposed change in a separate PR since it will require DOM change.

philn replied:
Also this introduced a warning here:
```
Source/WebCore/css/html.css:862:24: warning: empty character constant [-Winvalid-pp-token]
  862 |     content: '\2713' / '';
      |                        ^
1 warning generated.
```

fantasai replied:
@philn Empty strings in CSS have a meaning that is distinct, why is it generating a warning?

nt1m replied:
Chris fixed this by using double quotes :)

nt1m replied:
https://commits.webkit.org/306998@main

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -847,6 +856,31 @@ input:is([type="checkbox"], [type="radio"]):checked:disabled {
+    content: '\2713' / '';
+    height: 100%;
+    width: 100%;
+    display: inline-block;
```
The spec currently defines the input as `display: flex` which I'm not seeing here. That would also mean there's no reason to set `display: inline-block` here.

shallawa replied:
The [specs](https://www.w3.org/TR/css-forms-1/#stylesheet-checkbox-radio) state that the display of the radio checkmark should be `display: inline-block;`
```
input[type=radio]:checked::checkmark {
    background-color: currentColor;
    display: inline-block;
    border-radius: inherit;
    height: 100%;
    width: 100%;
}
```

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -847,6 +856,31 @@ input:is([type="checkbox"], [type="radio"]):checked:disabled {
+    height: 100%;
+    width: 100%;
+    display: inline-block;
+    background-color: currentColor;
```
Wouldn't this make the checkmark invisible?

shallawa replied:
I think `currentColor` inherits the `color` of the checkbox. In the layout tests, I did the following:
```
    input:is([type=checkbox]:not([switch]), [type=radio]) {
        color: darkgreen;
    }

    input[type=checkbox]:not([switch]):checked::checkmark {
        background-color: currentColor;
        mask: url("data:image/svg+xml, ....);
    }
```
So the `checkmark` has `currentColor` from the `color` of checkbox which is `darkgreen`. And since it has an SVG mask, this will draw only the SVG image colored by `darkgreen`.

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -847,6 +856,31 @@ input:is([type="checkbox"], [type="radio"]):checked:disabled {
+}
+
+input:is([type=checkbox]:not([switch]), [type=radio]):not(:checked)::checkmark {
+    display: none;
```
See above comment wrt `visibility`.

fantasai commented on Source/WebCore/css/html.css:
```diff
@@ -847,6 +856,31 @@ input:is([type="checkbox"], [type="radio"]):checked:disabled {
+}
+
+input:is([type=checkbox]:not([switch]), [type=radio]):checked:disabled::checkmark {
+    background-color: color-mix(in srgb, currentColor 30%, transparent);
```
Doesn't match [the spec](https://drafts.csswg.org/css-forms/#basic-appearance-stylesheet). Presumably the spec needs updating here, are there any outstanding decisions that need to be edited in or do we need to file an issue?

shallawa replied:
This was removed. I will follow up with @annevk and @nt1m about what should be added to the checkbox and radio button in the UA sheet.

lukewarlow replied:
https://github.com/w3c/csswg-drafts/issues/12225 - relevant spec issue. We need to actually decide on what styles we want here.

annevk reviewed (approved):
The Linux bot failures look legitimate, but should be simple rebasing of test expectations I think.

shallawa committed:
[appearance: base] Add default `base` properties for checkbox and radio button in UA style sheet
https://bugs.webkit.org/show_bug.cgi?id=304092
rdar://166415728

Reviewed by Anne van Kesteren.

-internal-auto-base() can only be parsed in the UA style sheet. Adding conditional
properties to html.css will allow setting some of the CSS properties for checkbox
and radio button when they are styled with "appearance: base".

Remove all `-internal-auto-base()` properties from the existing layout tests. No
need anymore to enable `CSSInternalAutoBaseParsingEnabled` in these layout tests.

Specs: https://www.w3.org/TR/css-forms-1/#stylesheet-checkbox-radio

* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics.html:
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic-expected.html:
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-dynamic.html:
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-checkmark-initial.html:
* Source/WebCore/css/html.css:
(input:is([type="radio"], [type="checkbox"])):
(input[type="radio"]):
(input[type=checkbox]:not([switch]):checked::checkmark):
(input[type=radio]:checked::checkmark):
(input:is([type=checkbox]:not([switch]), [type=radio]):not(:checked)::checkmark):

Canonical link: https://commits.webkit.org/306540@main
